src/docmanager/index.ts

```diff
@@ -0,0 +1,948 @@
+   */
+  listWidgetFactories(path?: string): string[] {
+    path = path || '';
+    let ext = '.' + path.split('.').pop();
+    let factories: string[] = [];
+    let options: Private.IWidgetFactoryEx;
+    let name = '';
+    // If an extension was given, filter by extension.
+    // Make sure the modelFactory is registered.
+    if (ext.length > 1) {
+      if (ext in this._defaultWidgetFactories) {
+        name = this._defaultWidgetFactories[ext];
+        options = this._widgetFactories[name];
+        if (options.modelName in this._modelFactories) {
+          factories.push(name);
+        }
+      }
+    }
+    // Add the rest of the valid widgetFactories that can open the path.
+    for (name in this._widgetFactories) {
+      if (factories.indexOf(name) !== -1) {
+        continue;
+      }
+      options = this._widgetFactories[name];
+      if (!(options.modelName in this._modelFactories)) {
+        continue;
+      }
+      let exts = options.fileExtensions;
+      if ((ext in exts) || ('.*' in exts)) {
+        factories.push(name);
+      }
+    }
+    // Add the default widget if it was not already added.
+    name = this._defaultWidgetFactory;
+    if (name && factories.indexOf(name) === -1) {
+      options = this._widgetFactories[name];
+      if (options.modelName in this._modelFactories) {
+        factories.push(name);
+      }
+    }
+    return factories;
+  }
+
+  /**
+   * Get the kernel preference.
+   */
+  getKernelPreference(path: string, widgetName: string): IKernelPreference {
+    let widgetFactoryEx = this._getWidgetFactoryEx(widgetName);
+    let modelFactoryEx = this._getModelFactoryEx(widgetName);
+    let language = modelFactoryEx.factory.preferredLanguage(path);
+    return {
+      language,
+      preferKernel: widgetFactoryEx.preferKernel,
+      canStartKernel: widgetFactoryEx.canStartKernel
+    }
+  }
+
+  /**
+   * Open a file and return the widget used to display the contents.
+   *
+   * @param path - The file path to open.
+   *
+   * @param widgetName - The name of the widget factory to use.
+   *
+   * @param kernel - An optional kernel name/id to override the default.
+   */
+  open(path: string, widgetName='default', kernel?: IKernelId): Widget {
+    let widget = new Widget();
+    let manager = this._contentsManager;
+    let mFactoryEx: Private.IModelFactoryEx;
+    if (widgetName !== 'default') {
+      mFactoryEx = this._getModelFactoryEx(widgetName);
+    } else {
+      widgetName = this.listWidgetFactories(path)[0];
+      mFactoryEx = this._getModelFactoryEx(widgetName);
+    }
+    let lang = mFactoryEx.factory.preferredLanguage(path);
+    let model: IDocumentModel;
+    let id = this._contextManager.findContext(path, mFactoryEx.name);
+    if (id) {
+      model = this._contextManager.getModel(id);
+    } else {
+      model = mFactoryEx.factory.createNew(lang);
+    }
+    let opts = mFactoryEx.contentsOptions;
+    manager.get(path, opts).then(contents => {
```

Comment:
When the model exists already (i.e., the `if (id)` case above), this effectively reverts the model every time the file is opened. I think the fetch should only be done when we create a new model (in the `else` case above). The `this._createWidget` call should be outside of this just before the installMessageFilter. That may mean that the first widget creation will get an empty model which is later populated, but that's probably all right. It also means that if a model is dirty, opening up a new widget will reflect the dirty state instead of the state on disk, but that's probably all right too. On widget creation, the widget should indicate the dirty state if it is displaying an existing dirty model.